### ISSUE TITLE
Difference between Tslearn TimeSeriesKMeans and Scikit-learn KMeans

### ISSUE THREAD

**Issue by frankl1** [tslearn-team/tslearn#353]
**Describe the question**
Hi everyone and thank you for making the tslearn library. It is very useful and the documentation is clear.
I run TimeSeriesKMeans from tslearn and KMeans from scikit-learn on the Chinatown dataset. I tried to use the same parameters so that the found clusters should be the same for both algorithm. My time series have the same length and I used _euclidean_ as metric in TimeSeriesKMeans. I also used the same initial clusters for both methods.   

For both methods, I set the number of clusters to be equal to the number of time series in the dataset. However there are clusters that are at the same location at the beginning of the algorithm. 

**To Reproduce**
Here is Colab notebook to reproduce the experiment
https://colab.research.google.com/drive/1nDu0qLASexots_uai1w5xNnNZgSVuvAM?usp=sharing

**Expected behavior**
I expected both method to find the same clusters. However while tslearn method was able to find clusters that I am more statisfied with, scikit-learn method returned the same initial clusters. I would like to know what is the difference between TimeSeriesKMeans using euclidean distance and scikit-learn bease KMeans.

It would also be very helpful if anyone can tell me why scikit-learn KMeans is not clustering my data with these settings.

**Environment (please complete the following information):**
 - tslearn version 0.5.0.5

Thanks in advance

**Comment by rtavenar** [comment]
Hi @frankl1 

Indeed, both tslearn's and sklearn's k-means methods should be similar if you use euclidean distance on the tslearn side.
However, what your experiment suggests is that they behave differently when empty clusters are found: while tslearn will try to reset cluster centers, what you observe suggests that sklearn stops the process.
To be honest, I think that both ways are fair, and I don't fully understand why you could be interested in running k-means with as many clusters as samples.

**Comment by frankl1** [comment]
Alright thank you. I was wondering why the result was different. But, I don't clearly understand what you mean by _resetting cluster centers_. Does it mean restarting with new set of random centers?
 
I agree that there are many clusters. I was trying an idea that not even clear to me. Basally, I was hoping that many clusters were going to be empty and few (actually two ) will not at the end of KMeans.

**Comment by rtavenar** [comment]
The problem is that kmeans alternates between two basic steps:
1. Assign series to their closest centroid
2. Update centroids as the barycenters of their assigned series

At step 2, if no series is assigned to the cluster, the barycenter cannot be computed.
This is why one has to do something about empty clusters in kmeans: either stop the optimization process or reset the cluster centers by any means.

**Comment by frankl1** [comment]
Exactly.

I am wondering what it would take to modify KMeans so it can start with relatively large number of clusters and automatically prune/merge clusters during its iterations until convergence:

1. Assign points/series to their closest centroids
2. For each centroid if forget_condition is met then remove this centroid from the centroid list
3. Update the remaining centroid and restart at 1. until convergence.

I think I can close this issue since my question has already been answered. However, we can continue this discussion about KMean improvement.

Thank you
